mimxrt: implementation of LED class

alphaFred opened the pull request:
Implemented LED class with rudimentary parts of pin module to support it. The LED class does not yet support setting an intensity. 

Added support for MIMXRT1020-EVK since this is the board I have
available for testing. Tested build for all three boards.

alphaFred committed:
mimxrt: implementation of LED class

Implemented LED class with rudimentary parts of pin module to support it.
The LED class does not yet support setting an intensity.

Added support for MIMXRT1020-EVK since this is the board I have
available for testing. Tested build for all three boards though.

jimmo commented on ports/mimxrt/modpyb.c:
```diff
@@ -0,0 +1,41 @@
+#include "py/runtime.h"
+#include "led.h"
+
+STATIC const mp_rom_map_elem_t pyb_module_globals_table[] = {
```
I'm not 100% sure that this should be in the "pyb" module. Some discussion probably required, but I see the pyb module as primarily "for the pyboard", and for historical reasons, also "STM32 functionality that hasn't made it to the machine module yet".

Unfortunately there isn't really a good precedent here, but I think this is a good an opportunity as ever to make a decision... LED is kind of a funny one...it's somewhere between machine and "board".

Perhaps we could defer this decision by just exposing the Pin object for now (i.e. machine.Pin), just providing the high/low/value methods.

@dpgeorge ?

dpgeorge replied:
I agree, there should not be a `pyb` module in this port.

The nrf port has a `board` module with `board.LED`.  And `examples/hwapi/` puts `LED` configuration in a Python `hwconfig` module.  But regardless of whether it's implemented in C or Python, the important decision is where to put LED, and whether it's just an alias of Pin/Signal, or is a specific implementation.

> Perhaps we could defer this decision by just exposing the Pin object for now (i.e. machine.Pin), just providing the high/low/value methods.

Yes, that would be the easy path.  Although it is nice to have pre-defined LEDs on the correct pins so you don't have to look it up in the board reference.

alphaFred replied:
I guess I could put it in the ```machine``` module then in addition to a future ```pin``` object. 
My take was more like looking at the STM32 port to allow for code reuse. Hence I put it into the ``` pyb``` module. But of course it makes not much sense to have this module in that particular port.

alphaFred replied:
done - moved to ```machine``` module and removed ```modpyb``` and any references to ```pyb``` from port

jimmo reviewed (commented):
Thanks @alphaFred, great to see this contribution!

Overall looks good, just some questions and comments inline.

I would be keen to see the same make-pins.py approach via a CSV file such that the functionality (both the Python API but also how board definitions are made) matches STM32, but I think this is a reasonable step towards that.

See inline comment about modpyb... some thought required there.

Ideally the two logical parts of this PR (adding the 1020 board and adding pin/led) would be two separate commits.

alphaFred commented:
> Overall looks good, just some questions and comments inline.
Thank you 😊 

> I would be keen to see the same make-pins.py approach via a CSV file such that the functionality (both the Python API but also how board definitions are made) matches STM32, but I think this is a reasonable step towards that.

Yes indeed. As a matter of fact I am currently working on that. My reasoning was to first introduce the LED module in a separate commit. Especially since it is my first contribution so I wanted to keep it somewhat limited in scope. 

> See inline comment about modpyb... some thought required there.
Will work through your comments. Thank you for the review. 

> Ideally the two logical parts of this PR (adding the 1020 board and adding pin/led) would be two separate commits.
Agreed,  I will split it up.

dpgeorge commented:
> My reasoning was to first introduce the LED module in a separate commit. Especially since it is my first contribution so I wanted to keep it somewhat limited in scope.

Yes, thank you, that is the best way to contribute here, and easiest way to get things merged: to make the changes small and self contained.

alphaFred commented on ports/mimxrt/boards/MIMXRT1020_EVK/evkmimxrt1020_flexspi_nor_config.h:
```diff
@@ -0,0 +1,268 @@
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
```
done - since PR has been split up into LED and MIMXRT1020_EVK integration changes will be visible in separate PR

alphaFred commented on ports/mimxrt/modpyb.c:
```diff
@@ -0,0 +1,41 @@
+const mp_obj_module_t mp_module_pyb = {
+    .base = {&mp_type_module},
+    .globals = (mp_obj_dict_t *)&pyb_module_globals,
+};
```
done - file removed after discussion

alphaFred commented on ports/mimxrt/pin.h:
```diff
@@ -0,0 +1,108 @@
+ */
+
+#ifndef MICROPY_INCLUDED_MIMXRT10XX_PIN_H
+#define MICROPY_INCLUDED_MIMXRT10XX_PIN_H
```
done - also fixed in led.h

alphaFred commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,196 @@
+
+
+void led_init(void) {
+    /* Turn off LEDs and initialize */
```
done - hopefully got all of them ;-)

alphaFred commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,196 @@
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Damien P. George
+ * Copyright (c) 2020 Jim Mussared
```
done

alphaFred commented:
Hi @dpgeorge 
the checks for my commit have now failed a couple of times because of reasons I believe are not related to my PR (e.g. code size increase in bare-arm port).

I am not too sure how to react on that now. Can I retrigger the checks (for instance in case of network related problems e.g. can't pull dependencies)?
Or shall I ignore the results and the PR will be reviewed and possible merged nonetheless?

Sorry for asking theses stupid questions :-)

best regards
alphaFred

dpgeorge commented:
> Or shall I ignore the results and the PR will be reviewed and possible merged nonetheless?

Just ignore that failure, it is spurious.

tve commented:
> > Or shall I ignore the results and the PR will be reviewed and possible merged nonetheless?
> 
> Just ignore that failure, it is spurious.

I'm afraid you're being generous ;-). That failure is systematic as far as I can tell, I see it in all the PRs I look at.

jimmo commented:
@alphaFred Thanks for making the changes, I think this looks good and provides a good base to start extending these features (even if it means some things move around or change).

If you have further ideas on things you'd like to add to this port, please feel free to open an issue or forum thread to discuss, otherwise these small contained PRs work well too.

FWIW, one of the TODOs on my list for this port (in addition to the obvious stuff like Pin, ADC, etc) was what appears to be a buffer issue with the USB CDC, which is currently preventing pyboard.py from working (which makes it hard to run the automated tests).

alphaFred commented:
> If you have further ideas on things you'd like to add to this port, please feel free to open an issue or forum thread to discuss, otherwise these small contained PRs work well too.

Ok, sure. 

> FWIW, one of the TODOs on my list for this port (in addition to the obvious stuff like Pin, ADC, etc) was what appears to be a buffer issue with the USB CDC, which is currently preventing pyboard.py from working (which makes it hard to run the automated tests).

Mhmh. This sounds like something that should be fixed 😊.  Is there already an issue/ forum thread about it? I could not find anything 🤔.

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.h:
```diff
@@ -3,6 +3,7 @@
-#define MICROPY_HW_LED_PORT    GPIO1
-#define MICROPY_HW_LED_PIN     11
+// i.MX RT1010 EVK has 1 board LED
+#define MICROPY_HW_LED_PIN (GPIO_11)
```
perhaps this should be `MICROPY_HW_LED1_PIN` to support multiple LEDs in the future?

dpgeorge commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,193 @@
+/// \moduleref machine
+/// \class LED - LED object
+///
+/// The LED object controls an individual LED (Light Emitting Diode).
```
These triple-slash quotes are left over from long ago when the docs were automatically extracted from the C code, and are no longer needed, nor are the meta tags like `\moduleref` and `\class`.  I understand this was copied from code in stm32 and that should be removed there.  But for now please just use normal `//` comments and none of the meta data.

alphaFred replied:
Yes, I took stm32 as a reference. Will remove these kind of comments 👌

dpgeorge commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,193 @@
+
+
+/******************************************************************************/
+/* MicroPython bindings                                                       */
```
It would be good to split out all the code below to a separate file `machine_led.c` which covers just the bindings C<->Python.

dpgeorge commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,193 @@
+/// \classmethod \constructor(id)
+/// Create an LED object associated with the given LED:
+///
+///   - `id` is the LED number, 1.
```
no need for the triple-slash or meta data

dpgeorge commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,193 @@
+/// If an argument is given, set the LED intensity and return `None`.
+mp_obj_t led_obj_intensity(size_t n_args, const mp_obj_t *args) {
+    return mp_const_none;
+}
```
I'd say just leave this unimplemented function out for now.  That way the code raises an exception when trying to access it, rather than failing silently.

dpgeorge commented on ports/mimxrt/led.c:
```diff
@@ -0,0 +1,193 @@
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(led_obj_on_obj, led_obj_on);
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(led_obj_off_obj, led_obj_off);
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(led_obj_toggle_obj, led_obj_toggle);
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(led_obj_intensity_obj, 1, 2, led_obj_intensity);
```
This defines should be split up and moved to just below their respective function definitions.

dpgeorge commented on ports/mimxrt/main.c:
```diff
@@ -41,6 +42,10 @@ void board_init(void);
+    led_init();
+
+    // Enable board LED after initialization
+    led_state(MACHINE_BOARD_LED, 1U);
```
I think it best not to turn on LEDs by default but rather leave them off.  That's how the stm32 boards work.

dpgeorge commented on ports/mimxrt/modmachine.c:
```diff
@@ -48,6 +50,9 @@ STATIC const mp_rom_map_elem_t machine_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_mem32),               MP_ROM_PTR(&machine_mem32_obj) },
+    { MP_ROM_QSTR(MP_QSTR_LED),                 MP_ROM_PTR(&machine_led_type) },
+    // class constants
+    { MP_ROM_QSTR(MP_QSTR_BOARD_LED_ID),        MP_ROM_INT(MACHINE_BOARD_LED) },
```
I don't think this is needed.  Most peripherals (eg Pin, I2C, SPI, UART) require you to know what value to pass to the constructor to get the right peripheral.  Rather it should be part of the documentation for each board.

alphaFred committed:
implemented review findings

yanyao12138 commented:
Hi @alphaFred, I have implemented pin class/led class and make-pin.py locally, maybe we can discuss this topic and try to integrate our work.😊

alphaFred commented:
> Hi @alphaFred, I have implemented pin class/led class and make-pin.py locally, maybe we can discuss this topic and try to integrate our work.😊

Hi. Sounds great. Do you have your changes accessible via github so that I can take a look? 
@dpgeorge   for the LED class I hope I have fixed now all findings and it can be merged soon 😊. But especially for pin module I am struggling a bit so having someone to discuss this is very much appreciated 😄.

dpgeorge commented:
@alphaFred ahh, sorry, I don't think I got any notification that you force-pushed the branch with changes.  Will look at it ASAP.

alphaFred commented:
> @alphaFred ahh, sorry, I don't think I got any notification that you force-pushed the branch with changes.  Will look at it ASAP.

Cool. No worries. I just thought I kindly remind you 😊. Looking forward to it.

yanyao12138 commented:
> > Hi @alphaFred, I have implemented pin class/led class and make-pin.py locally, maybe we can discuss this topic and try to integrate our work.😊
> 
> Hi. Sounds great. Do you have your changes accessible via github so that I can take a look?
> @dpgeorge for the LED class I hope I have fixed now all findings and it can be merged soon 😊. But especially for pin module I am struggling a bit so having someone to discuss this is very much appreciated 😄.

Hi @alphaFred , you can find my code in this [link](https://github.com/yanyao12138/micropython/commit/60b6ac1649d835b15e65e9c9c185039b75ab6a07). I also created the make-pins.py to interpret the .csv file. For the pin class, I think there are still someplace to be optimized. I hope we can discuss this 😊. Thanks for your support of NXP mimxrt port.

dpgeorge commented:
Ok, this has now been rebased, squashed and merged in 02cc4462b70729a42cb127c840fee891dd08a0d4, with some minor changes to get it building on MIMXRT1060_EVK which does not have any LEDs defined.

Thanks @alphaFred for the work and contribution!

MaureenHelm commented:
> Ok, this has now been rebased, squashed and merged in [02cc446](https://github.com/micropython/micropython/commit/02cc4462b70729a42cb127c840fee891dd08a0d4), with some minor changes to get it building on MIMXRT1060_EVK which does not have any LEDs defined.

FYI @yanyao12138

yanyao12138 commented:
> > Ok, this has now been rebased, squashed and merged in [02cc446](https://github.com/micropython/micropython/commit/02cc4462b70729a42cb127c840fee891dd08a0d4), with some minor changes to get it building on MIMXRT1060_EVK which does not have any LEDs defined.
> 
> FYI @yanyao12138

Got it, I have created the PR.
